Feature/lab 1

starfallz opened the pull request:
Fixes #428 

Review of colleague's PR #427 

#### Changes proposed in this PR:

- My very first PR for go yay! Fib function that takes both positive / negative values with tests included.

hamana-anz committed:
LAB_1: added main that calls fib function with tests.

the fib function should work with both positive and negative numbers.

hamana-anz committed:
LAB_1: modified code and test due to linting issues.

Changed the array equal using deepEqual instead of individually checking for expected values.

hamana-anz committed:
LAB_1: modified code and test due to linting issues.

- Changed the array equal using reflect.deepEqual instead of individually checking for expected values.
- Fixed some minor spacing issues.

juliaogris commented on 01_fib/starfallz/main.go:
```diff
@@ -0,0 +1,44 @@
+var out io.Writer = os.Stdout
+
+func main() {
+	fmt.Fprint(out, strings.Trim(strings.Join(strings.Split(fmt.Sprint(fib(7)), " "), "\n"), "[]"))
```
That's a novel approach :)

However, the lab requirement was to write a function with the following signature:  
`func fib(n int)` 
happy to let this go.

You could pull the `Split` and `Join` together with a `strings.ReplaceAll`.
I agree it's nice that you don't print inside `fib`, however, I'd wrap the printing in a `printIntSlice` function or similar, that iterates through the `fib`'s result array `Println`s each element. 
In order to understand your approach the reader of your code has to have intimate knowledge with the way golang prints slices.

starfallz replied:
Yeah initially I did that however it felt kind of wierd to do many things in one function. I will implement the pritntIntSlice func instead since I want to be able to test them separately and not do too much in one func

juliaogris replied:
Sounds good!

juliaogris commented on 01_fib/starfallz/main.go:
```diff
@@ -0,0 +1,44 @@
+		n *= -1
+	}
+
+	result := []int{1}
```
`result := make([]int, n)` - you already know how big you need your slice to be.
slightly shorter:
```go
	sign := 1
	if n < 0 {
		sign = -1
	}

	n1, n2 := 0, 1
	result := make([]int, n*sign)
	for i := 0; i < n*sign; i++ {
		n1, n2 = n2, n1+sign*n2
		result[i] = n1
	}
	return result
}
```

if you print inside the function as required you should get a shortened:

```go
	sign := 1
	if n < 0 {
		sign = -1
	}

	n1, n2 := 0, 1
	for j := 0; j < n*sign; j++ {
		n1, n2 = n2, n1+sign*n2
		fmt.Fprintln(out, n1)
	}
```

starfallz replied:
Haha you are literally giving me the solution Julia 😂😂😂😂 I agree it's much shorter but should I be copy pasting your code? Lol feels like it's a completely new implementation haha

juliaogris replied:
I know - can't help myself ;)

**TL;DR: You don't need to update anything (but you certainly can if you want ;) ).**

Some of us running this course have spent a bit of time going back and forth between us on what we think the canonical solution should be, and I always find it interesting to see how different the same short piece of functionality can look.  

Personally, I love it when others look at my code and help me make it shorter, nicer and/or more readable. Everything that makes it better makes _me_  happy: removing little stumbling blocks like typos, bad variable names, going against conventions and idioms or poor naming. Even more important I really appreciate it when medium size weirdnesses on a block/logic level are clarified or even bigger poor design choices are brought to light. 

To me it's not just about getting the job/ticket done but to have fun. To achieve this I see my job as a craft that wants to express itself as beautifully as it can and it almost always becomes better with many eyes looking over it. That's why I posted it :)

I guess that was the long answer.

juliaogris commented on 01_fib/starfallz/main_test.go:
```diff
@@ -0,0 +1,59 @@
+		input       int
+		expected    []int
+	}{
+		{"Test zero fib to return zero", 0, []int{0}},
```
Arguably this shouldn't print anything. Up to you.

starfallz replied:
Ok i will return empty array for this instead of 0

juliaogris commented on 01_fib/starfallz/main_test.go:
```diff
@@ -0,0 +1,59 @@
+		t.Run(testCase.description, func(t *testing.T) {
+			actual := fib(input)
+
+			if !reflect.DeepEqual(actual, expected) {
```
Might be worth while looking into `testify/assert` and `assert.Equal`, it pulls these three lines into one and prints nice diff in case of failure.

starfallz replied:
Initially I wanted to explore testify but I looked at lab6 and that's where it specifically ask us to use the assert functions. So I kept it plain vanilla go so I know how to do it without external dependency. But I'm happy to modify it to assert equal if it's a requirement

juliaogris replied:
ok.

juliaogris reviewed (changes requested):
Please [fix commit messages](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#commit-messages) with `git rebase -i`

hamana-anz committed:
LAB_1: modified code and test based on PR comments.

- I have modified the main to call printIntSlice(fib(7)) instead of the confusingly long join splits.
- I have added new printIntSlice function that prints int with new line for every int.
- Added tests for printIntSlice
- Modified fib(0) to return []
- Modified fib function to be more robust and simple based on comment.

hamana-anz committed:
LAB_1: modified code and test based on PR comments.

- I have modified the main to call printIntSlice(fib(7)) instead of the confusingly long join splits.
- I have added new printIntSlice function that prints int with new line for every int.
- Added tests for printIntSlice
- Modified fib(0) to return []
- Modified fib function to be more robust and simple based on comment.

hamana-anz committed:
Merge branch 'feature/lab_1' of https://github.com/starfallz/go-course into feature/lab_1

juliaogris reviewed (approved):
Code LGTM.
Usually I'd ask for a git commit clean up, but I don't want to block you any more.
Please have a read through the [git commit message standards](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#commit-messages) we are asking for.
It's invaluable to learn the usage of `git rebase -i` to re-write and clean up you commit history. If you do want to fix up your commits before merge, please reach out on Slack and I will re-approve right away.

starfallz commented:
@juliaogris i tried rebasing via sourcetree, (kinda screwed things up) then i tried manually, kinda worked, but my biggest issue is my initial commit, that cant be changed so i gave up hahaha, if you are in the office can u show me how to reset the initial commit msg? i heard i can use --root but i havent actually tried it.

juliaogris commented:
The go-course is now closed. Thank you very much for participating.
